File: leo2-bridge-com/LeoBridge/Service/RoutingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ServiceModel;
using System.ServiceModel.Web;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using System.Runtime.InteropServices;

namespace LeoBridge.Service
{
    [ServiceContract]
    [ComVisible(true)]
    public interface IRoutingService
    {
        [OperationContract]
        [WebInvoke(
            RequestFormat = WebMessageFormat.Json,
            ResponseFormat = WebMessageFormat.Json,
            UriTemplate = "/rs/api/v1/routing/request",
            Method = "POST"
            )]
        Routing Request(RoutingRequest r);
    }

    #region RoutingRequest

    #region RoutingRequestParticipant
    [ComVisible(true)]
    public interface IRoutingRequestParticipant
    {
        string Country { get; set; }
        string Zip { get; set; }
        string TimeFrom { get; set; }
        string TimeTo { get; set; }
        string DesiredStation { get; set; }
    }

    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]
    [DataContract]
    public class RoutingRequestParticipant : IRoutingRequestParticipant
    {
        public RoutingRequestParticipant() { }

        [DataMember(Name = "country", IsRequired = false, EmitDefaultValue = false)]
        public string Country { get; set; }
        [DataMember(Name = "zip", IsRequired = false, EmitDefaultValue = false)]
        public string Zip { get; set; }
        [DataMember(Name = "timeFrom", IsRequired = false, EmitDefaultValue = false)]
        public string TimeFrom { get; set; }
        [DataMember(Name = "timeTo", IsRequired = false, EmitDefaultValue = false)]
        public string TimeTo { get; set; }
        [DataMember(Name = "desiredStation", IsRequired = false, EmitDefaultValue = false)]
        public string DesiredStation { get; set; }
    }
    #endregion

    [ComVisible(true)]
    public interface IRoutingRequest
    {
        String SendDate { get; set; }
        String DesiredDeliveryDate { get; set; }
        RoutingRequestParticipant Sender { get; set; }
        RoutingRequestParticipant Consignee { get; set; }
        int Services { get; set; }
        double Weight { get; set; }
    }

    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]
    [DataContract]
    public class RoutingRequest : IRoutingRequest
    {
        public RoutingRequest() { }

        [DataMember(Name = "sendDate", IsRequired = true)]
        public String SendDate { get; set; }
        [DataMember(Name = "desiredDeliveryDate", EmitDefaultValue=false)]
        public String DesiredDeliveryDate { get; set; }
        [DataMember(Name = "sender")]
        public RoutingRequestParticipant Sender { get; set; }
        [DataMember(Name = "consignee")]
        public RoutingRequestParticipant Consignee { get; set; }
        [DataMember(Name = "services")]
        public int Services { get; set; }
        [DataMember(Name = "weight")]
        public double Weight { get; set; }
    }
    #endregion

    #region Routing

    #region RoutingParticiapnt
    [ComVisible(true)]
    public interface IRoutingParticipant
    {
        string Station { get; set; }
        string Country { get; set; }
        string ZipCode { get; set; }
        string Zone { get; set; }
        string Sector { get; set; }
        string DayType { get; set; }
        Boolean Island { get; set; }
        int Term { get; set; }
        string EarliestTimeOfDelivery { get; set; }
        string SaturdayDeliveryUntil { get; set; }
        string SundayDeliveryUntil { get; set; }
        string PickupUntil { get; set; }
        string PartnerManager { get; set; }
        string ToString();
    }

    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]
    [DataContract]
    public class RoutingParticipant : IRoutingParticipant
    {
        public RoutingParticipant() { }

        [DataMember(Name = "station", IsRequired = false, EmitDefaultValue = false)]
        public string Station { get; set; }
        [DataMember(Name = "country", IsRequired = false, EmitDefaultValue = false)]
        public string Country { get; set; }
        [DataMember(Name = "zipCode", IsRequired = false, EmitDefaultValue = false)]
        public string ZipCode { get; set; }
        [DataMember(Name = "zone", IsRequired = false, EmitDefaultValue = false)]
        public string Zone { get; set; }
        [DataMember(Name = "sector", IsRequired = false, EmitDefaultValue = false)]
        public string Sector { get; set; }
        [DataMember(Name = "dayType", IsRequired = false, EmitDefaultValue = false)]
        public string DayType { get; set; }
        [DataMember(Name = "island", IsRequired = false)]
        public Boolean Island { get; set; }
        [DataMember(Name = "term", IsRequired = false)]
        public int Term { get; set; }
        [DataMember(Name = "earliestTimeOfDelivery", IsRequired = false, EmitDefaultValue = false)]
        public string EarliestTimeOfDelivery { get; set; }
        [DataMember(Name = "saturdayDeliveryUntil", IsRequired = false, EmitDefaultValue = false)]
        public string SaturdayDeliveryUntil { get; set; }
        [DataMember(Name = "sundayDeliveryUntil", IsRequired = false, EmitDefaultValue = false)]
        public string SundayDeliveryUntil { get; set; }
        [DataMember(Name = "pickupUntil", IsRequired = false, EmitDefaultValue = false)]
        public string PickupUntil { get; set; }
        [DataMember(Name = "partnerManager", IsRequired = false, EmitDefaultValue = false)]
        public string PartnerManager { get; set; }

        public override string ToString()
        {
            return String.Format("Station [{0}] Country [{1}] ZipCode [{2}] Zone [{3}] Sector [{4}] Day type [{5}] Island [{6}] Term [{7}]",
                this.Station,
                this.Country,
                this.ZipCode,
                this.Zone,
                this.Sector,
                this.DayType,
                this.Island,
                this.Term);
        }
    }
    #endregion

    [ComVisible(true)]
    public interface IRouting
    {
        String SendDate { get; set; }
        String DeliveryDate { get; set; }
        RoutingParticipant Sender { get; set; }
        RoutingParticipant Consignee { get; set; }
        String[] ViaHubs { get; set; }
        String LabelContent { get; set; }
        String Message { get; set; }

        Error Error { get; set; }
        String ToString();
    }

    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]
    [DataContract]
    public class Routing : ErrorContainer, IRouting
    {
        public Routing() { }

        [DataMember(Name = "sendDate", IsRequired = true)]
        public String SendDate { get; set; }

        [DataMember(Name = "deliveryDate", IsRequired = false)]
        public String DeliveryDate { get; set; }

        [DataMember(Name = "sender", IsRequired = false)]
        public RoutingParticipant Sender { get; set; }

        [DataMember(Name = "consignee", IsRequired = false)]
        public RoutingParticipant Consignee { get; set; }

        [DataMember(Name = "viaHubs", IsRequired = false)]
        public String[] ViaHubs { get; set; }

        [DataMember(Name = "labelContent", IsRequired = false)]
        public String LabelContent { get; set; }

        [DataMember(Name = "message", IsRequired = false)]
        public String Message { get; set; }

        public override string ToString()
        {
            return String.Format("Send date [{0}] Delivery date [{1}] Sender [{2}] Consignee [{3}] ViaHubs [{4}] Label content [{5}] Message [{6}]",
                this.SendDate,
                this.DeliveryDate,
                this.Sender,
                this.Consignee,
                this.ViaHubs,
                this.LabelContent,
                this.Message);
        }
    }
    #endregion

    #region Proxy
    /// <summary>
    /// COM proxy class for routing service
    /// </summary>
    public class RoutingServiceProxy : ServiceClientProxy<IRoutingService>, IRoutingService
    {
        public RoutingServiceProxy(ChannelFactory<IRoutingService> factory)
            : base(factory) { }

        public Routing Request(RoutingRequest r)
        {
            return this.Call<Routing>(() => this.ServiceClient.Request(r) );
        }
    }
    #endregion
}

File: leoz-common/src/main/java/org/deku/leoz/service/pub/RoutingService.kt
package org.deku.leoz.service.pub

import io.swagger.annotations.*
import org.deku.leoz.service.entity.DayType
import org.deku.leoz.service.entity.ServiceError
import org.deku.leoz.service.entity.ShortDate
import org.deku.leoz.service.entity.ShortTime
import sx.rs.auth.ApiKey
import javax.ws.rs.Consumes
import javax.ws.rs.POST
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType

/**
 * Created by masc on 17.09.14.
 */
@Path("v1/routing")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Api(value = "Routing operations")
@ApiKey(false)
interface RoutingService {
    /**
     * Routing service specific error codes
     */
    enum class ErrorCode(private val value: Int) {
        ROUTE_NOT_AVAILABLE_FOR_GIVEN_PARAMETER(1000)
    }

    /**
     * Created by masc on 23.06.15.
     */
    @ApiModel(value = "RoutingRequest", subTypes = [(Request.Participant::class)], description = "Routing request")
    class Request {
        /**
         * Sender or consignee attributes
         * Created by masc on 23.06.15.
         */
        @ApiModel(value = "RequestParticipant", description = "Request participant. Sender or consignee")
        class Participant {
            @ApiModelProperty(dataType = "string", example = "DE", value = "Country two-letter ISO-3166", position = 10, required = true)
            var country: String? = null

            @ApiModelProperty(dataType = "string", example = "36286", value = "Zip code accordant to country specification", position = 20, required = true)
            var zip: String? = null

            @ApiModelProperty(dataType = "string", example = "09:00", position = 40, required = false, value = "Time window (from)", allowableValues = "00:00 - 23:59")
            var timeFrom: String? = null

            @ApiModelProperty(dataType = "string", example = "12:00", position = 50, required = false, value = "Time window (to)", allowableValues = "00:00 - 23:59")
            var timeTo: String? = null

            @ApiModelProperty(dataType = "string", example = "020", position = 60, required = false, value = "Desired station number", allowableValues = "010 - 999")
            var desiredStation: String? = null

            constructor() {}

            constructor(country: String? = null, zip: String? = null, timeFrom: String? = null, timeTo: String? = null, desiredStation: String? = null) {
                this.country = country
                this.zip = zip
                this.timeFrom = timeFrom
                this.timeTo = timeTo
                this.desiredStation = desiredStation
            }
        }

        @ApiModelProperty(dataType = "date", example = "2017-06-01", position = 10, required = true, value = "Send date", allowableValues = "2017-06-01")
        var sendDate: ShortDate? = null

        @ApiModelProperty(dataType = "date", example = "2017-06-02", position = 20, required = false, value = "Desired delivery date", allowableValues = "2017-06-02")
        var desiredDeliveryDate: ShortDate? = null

        @ApiModelProperty(value = "Sum of DeKu servic values", position = 22, required = false)
        var services: Int? = null

        @ApiModelProperty(value = "Real weight", position = 24, required = false)
        var weight: Float? = null

        @ApiModelProperty(value = "Sender", position = 30, required = false)
        var sender: Participant? = null

        @ApiModelProperty(value = "Consignee", position = 40, required = false)
        var consignee: Participant? = null

        @JvmOverloads constructor(
                sendDate: ShortDate? = null,
                desiredDeliveryDate: ShortDate? = null,
                services: Int? = null,
                weight: Float? = null,
                sender: Participant? = null,
                consignee: Participant? = null) {

            this.sendDate = sendDate
            this.desiredDeliveryDate = desiredDeliveryDate
            this.services = services
            this.weight = weight
            this.sender = sender
            this.consignee = consignee
        }
    }

    /**
     * Routing service request response
     * Created by JT on 23.06.15.
     */
    @ApiModel(value = "Routing", subTypes = [(Routing.Participant::class)], description = "Routing response")
    class Routing() {
        /**
         * Routing service request response member
         * Created by JT on 23.06.15.
         */
        @ApiModel(value = "Participant", description = "Response participant. Sender or consignee")
        class Participant {
            @ApiModelProperty(dataType = "string", example = "020", position = 10, required = true, value = "Station number", allowableValues = "010 - 999")
            var station: Int = 0

            val stationFormatted: String?
                get() = this.station.toString().padStart(3, '0')

            @ApiModelProperty(dataType = "string", example = "PL", position = 20, required = true, value = "Country two-letter ISO-3166")
            var country: String = ""

            @ApiModelProperty(dataType = "string", example = "01-1010", position = 30, required = true, value = "Zipcode contry conform")
            var zipCode: String = ""

            @ApiModelProperty(dataType = "string", example = "A", position = 40, required = true, value = "Zone", allowableValues = "A,B,C,D")
            var zone: String = ""

            @ApiModelProperty(dataType = "string", example = "WR", position = 45, required = false, value = "National zone", allowableValues = "WR,UL")
            var nationalZone: String = ""

            @ApiModelProperty(dataType = "string", example = "X", position = 50, required = true, value = "Sector", allowableValues = "A-Z")
            var sector: String = ""

            @ApiModelProperty(dataType = "string", example = "Workday", position = 60, required = true, value = "Type of day")
            var dayType: String = DayType.Workday.toString()

            @ApiModelProperty(example = "false", position = 70, required = true, value = "Is island")
            var island: Boolean = false

            @ApiModelProperty(dataType = "integer", example = "1", position = 80, required = true, value = "Termtime in days", allowableValues = ">=1")
            var term: Int = 1

            @ApiModelProperty(dataType = "string", example = "08:01", position = 90, required = true, value = "Earliest time of delivery", allowableValues = "00:00 - 23:59")
            var earliestTimeOfDelivery: ShortTime? = ShortTime()

            @ApiModelProperty(dataType = "string", example = "12:00", position = 120, required = true, value = "Delivery time until on saturday", allowableValues = "00:00 - 23:59")
            var saturdayDeliveryUntil: ShortTime = ShortTime(localTime = "00:00")

            @ApiModelProperty(dataType = "string", example = "12:00", position = 130, required = true, value = "Delivery time until on sunday", allowableValues = "00:00 - 23:59")
            var sundayDeliveryUntil: ShortTime = ShortTime(localTime = "00:00")

            @ApiModelProperty(dataType = "string", example = "16:00", position = 140, required = true, value = "Pick up time until", allowableValues = "00:00 - 23:59")
            var pickupUntil: ShortTime = ShortTime()

            @ApiModelProperty(dataType = "string", example = "AH", position = 150, required = true, value = "Partner manager")
            var partnerManager: String = ""

            @ApiModelProperty(hidden = true)
            var date: java.util.Date? = null

            @ApiModelProperty(hidden = true)
            var message: String? = ""
        }

        @ApiModelProperty(position = 30)
        var sender: Participant? = Participant()

        @ApiModelProperty(position = 40)
        var consignee: Participant? = Participant()

        @ApiModelProperty(dataType = "string", example = "F,N", position = 50, required = true, value = "Used via hubs: \"F,N\"")
        var viaHubs: Array<String>? = null

        @ApiModelProperty(dataType = "string", example = "F,N 100", position = 60, required = true, value = "Routing label content:  \"F,N 100\"")
        var labelContent: String = ""

        @ApiModelProperty(dataType = "string", example = "OK", position = 70, required = true, value = "Info message: \"OK\"")
        var message: String = ""

        @ApiModelProperty(dataType = "date", example = "2017-06-01", position = 10, required = true, value = "Send date", allowableValues = "2017-06-01")
        var sendDate: ShortDate? = null

        @ApiModelProperty(dataType = "date", example = "2017-06-02", position = 20, required = false, value = "Delivery date", allowableValues = "2017-06-01")
        var deliveryDate: ShortDate? = null
    }

    @POST
    @Path("/request")
    @ApiOperation(value = "Request routing information")
    @ApiResponses(
            ApiResponse(code = 400, message = "Bad request/parameter", response = ServiceError::class)
    )
    fun request(@ApiParam(value = "Routing request") routingRequest: Request): Routing
}
